File: walletcore/src/main/java/com/mycelium/wapi/wallet/TransactionSummary.java
package com.mycelium.wapi.wallet;

import com.google.common.base.Optional;
import com.mrd.bitlib.util.HexUtils;
import com.mycelium.wapi.wallet.coins.CryptoCurrency;
import com.mycelium.wapi.wallet.coins.AssetInfo;
import com.mycelium.wapi.wallet.coins.Value;

import org.jetbrains.annotations.NotNull;
import java.io.Serializable;
import java.util.List;

import javax.annotation.Nullable;

public class TransactionSummary implements Serializable, Comparable<TransactionSummary> {

    protected CryptoCurrency type;
    protected byte[] id;
    protected byte[] hash;
    protected Value transferred;
    protected long timestamp;
    protected List<InputViewModel> inputs;
    protected List<OutputViewModel> outputs;
    protected List<Address> destinationAddresses;
    protected int height;
    protected int confirmations;
    protected int rawSize;
    protected boolean isQueuedOutgoing;
    transient public Optional<ConfirmationRiskProfileLocal> confirmationRiskProfile;
    @Nullable
    protected Value fee;


    public TransactionSummary(CryptoCurrency type,
                                     byte[] id, byte[] hash,
                                     Value transferred,
                                     long timestamp,
                                     int height,
                                     int confirmations,
                                     boolean isQueuedOutgoing,
                              List<InputViewModel> inputs,
                              List<OutputViewModel> outputs,
                              List<Address> destinationAddresses,
                                     ConfirmationRiskProfileLocal risk,
                                     int rawSize, @Nullable Value fee) {
        this.type = type;
        this.id = id;
        this.hash = hash;
        this.transferred = transferred;
        this.timestamp = timestamp;
        this.confirmations = confirmations;
        this.height = height;
        this.isQueuedOutgoing = isQueuedOutgoing;
        this.inputs = inputs;
        this.outputs = outputs;
        this.destinationAddresses = destinationAddresses;
        this.confirmationRiskProfile = Optional.fromNullable(risk);
        this.rawSize = rawSize;
        this.fee = fee;
    }

    public boolean isQueuedOutgoing() {
        return isQueuedOutgoing;
    }

    public AssetInfo getType() {
        return type;
    }

    public int getConfirmations() {
        return confirmations;
    }

    public int getHeight() {
        return height;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    @Nullable
    public Value getFee() {
        return fee;
    }

    public List<InputViewModel> getInputs() {
        return inputs;
    }

    public List<OutputViewModel> getOutputs() {
        return outputs;
    }

    public List<Address> getDestinationAddresses() {
        return destinationAddresses;
    }

    public Value getTransferred() {
        return transferred;
    }

    public boolean isIncoming() {
        return transferred.moreOrEqualThanZero();
    }

    public int getRawSize() {
        return rawSize;
    }

    public byte[] getId() {
        return id;
    }

    public String getIdHex() {
        return HexUtils.toHex(id);
    }

    public Optional<ConfirmationRiskProfileLocal> getConfirmationRiskProfile() {
        return confirmationRiskProfile;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TransactionSummary other = (TransactionSummary) o;
        return getId().equals(other.getId());
    }

    @Override
    public String toString(){
        return getIdHex();
    }

    public boolean canCancel() {
        return isQueuedOutgoing;
    }

    @Override
    public int compareTo(@NotNull TransactionSummary other) {
        // TODO: Fix block heights! Currently the block heights are calculated as latest block height - confirmations + 1 but as it's not atomically collecting all the data, we run off by one frequently for transactions that get synced during a block being discovered.

        // Blockchains core property is that they determine the sorting of transactions.
        // In Bitcoin, timestamps of transactions are not required to be increasing, so the
        // block height is what has to be sorted by for transactions from different blocks.
        // if (other.getHeight() != getHeight()) {
        //     return other.getHeight() - getHeight();
        // }

        // If no block height is available (alt coins?), we do sort by timestamp.
        if (other.getTimestamp() != getTimestamp()) {
            return (int) (other.getTimestamp() - getTimestamp());
        }
        // Transactions are sorted within a block, too. Here we don't have that sequence number
        // handy but to ensure stable sorting, we have to sort by something robust.
        return getIdHex().compareTo(other.getIdHex());
    }
}

File: walletcore/src/main/java/com/mycelium/wapi/wallet/TransactionSummary.kt
package com.mycelium.wapi.wallet

import com.google.common.base.Optional
import com.mycelium.wapi.wallet.coins.CryptoCurrency
import com.mycelium.wapi.wallet.coins.AssetInfo
import com.mrd.bitlib.util.HexUtils
import com.mycelium.wapi.wallet.coins.Value
import com.mycelium.wapi.wallet.eth.EthAccount
import java.io.Serializable

open class TransactionSummary(var type: CryptoCurrency,
                              var id: ByteArray, protected var hash: ByteArray,
                              var transferred: Value,
                              var timestamp: Long,
                              var height: Int,
                              var confirmations: Int,
                              var isQueuedOutgoing: Boolean,
                              var inputs: List<InputViewModel>,
                              var outputs: List<OutputViewModel>,
                              var destinationAddresses: List<Address>,
                              risk: ConfirmationRiskProfileLocal?,
                              rawSize: Int, fee: Value?) : Serializable,
    Comparable<TransactionSummary> {
    var rawSize: Int
        protected set

    @Transient
    var confirmationRiskProfile: Optional<ConfirmationRiskProfileLocal> =
        Optional.fromNullable(risk)
    var fee: Value?
        protected set
    val isIncoming: Boolean
        get() = transferred.moreOrEqualThanZero()
    val idHex: String
        get() = HexUtils.toHex(id)

    init {
        this.rawSize = rawSize
        this.fee = fee
    }

//    fun getType(): AssetInfo {
//        return type
//    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val other = o as TransactionSummary
        return id.contentEquals(other.id)
    }

    override fun toString(): String {
        return idHex
    }

    fun canCancel(): Boolean {
        return isQueuedOutgoing
    }

    override fun compareTo(other: TransactionSummary): Int {
        // TODO: Fix block heights! Currently the block heights are calculated as latest block height - confirmations + 1 but as it's not atomically collecting all the data, we run off by one frequently for transactions that get synced during a block being discovered.

        // Blockchains core property is that they determine the sorting of transactions.
        // In Bitcoin, timestamps of transactions are not required to be increasing, so the
        // block height is what has to be sorted by for transactions from different blocks.
        // if (other.getHeight() != getHeight()) {
        //     return other.getHeight() - getHeight();
        // }

        // If no block height is available (alt coins?), we do sort by timestamp.
        return if (other.timestamp != timestamp) {
            (other.timestamp - timestamp).toInt()
        } else {
            // Transactions are sorted within a block, too. Here we don't have that sequence number
            // handy but to ensure stable sorting, we have to sort by something robust.
            idHex.compareTo(other.idHex)
        }
    }
}

fun TransactionSummary.isMinerFeeTx(account: WalletAccount<*>): Boolean =
    (this as? EthTransactionSummary)?.run {
        account is EthAccount && !isIncoming
                && value.isZero() && hasTokenTransfers
    } ?: false